flashlight/ext/integrations/halide/HalideInterface.cpp

```diff
@@ -0,0 +1,173 @@
+   if (res != CUDA_SUCCESS) throw std::runtime_error("cuCtxGetCurrent failed");
+   cudaStream_t stream = fl::cuda::getActiveStream();
+   fl::ext::detail::UserContext userCtx(deviceId, &ctx, &stream);
+   // This symbol is searched for by LLVM on the stack before
+   // JMPing to a function pointer
+   void* __user_context = (void*)&userCtx;
+
+ * \endcode
+ *
+ * I couldn't quite get this to work, but it might work better and not have some
+ * side effects that the current implementation does. Unclear.
+ */
+extern "C" {
+
+int halide_cuda_device_malloc(void* /* user_context */, halide_buffer_t* buf) {
+  size_t size = buf->size_in_bytes();
+  // TODO(jacobkahn): replace me with af::allocV2 when using AF >= 3.8
+  void* ptr = af::alloc(size, af::dtype::u8);
+  buf->device = (uint64_t)ptr; // eh
+  buf->device_interface = halide_cuda_device_interface();
+  // This doesn't work because the public device interface API returns an
+  // incomplete type. Is it needed? Who knows
+  // buf->device_interface->impl->use_module();
+  return 0;
+}
+
+int halide_cuda_device_free(void* /* user_context */, halide_buffer_t* buf) {
+  // TODO(jacobkahn): replace me with af::freeV2 when using AF >= 3.8
+  af::free((void*)buf->device);
+
+  // See above - we never call use_module(), so don't release it I suppose...
+  // buf->device_interface->impl->release_module();
+  buf->device_interface = nullptr;
+  buf->device = 0;
+  return 0;
+}
+
+int halide_cuda_acquire_context(
+    void* /* user_context */,
+    CUcontext* ctx,
+    bool create = true) {
+  CUcontext _ctx = 0;
+  CUresult res = cuCtxGetCurrent(&_ctx);
+  if (res != CUDA_SUCCESS) {
+    throw std::runtime_error("Could not get from CUDA context");
+  };
+  *ctx = _ctx;
+  return 0;
+}
+
+int halide_cuda_get_stream(
+    void* /* user_context */,
+    CUcontext /* ctx */,
+    CUstream* stream) {
+  *stream = (CUstream)fl::cuda::getActiveStream();
+  return 0;
+}
+
+int halide_get_gpu_device(void* /* user_context */) {
+  return af::getDevice();
+}
+
+} // extern "C"
+
+namespace fl {
+namespace ext {
+
+std::vector<int> afToHalideDims(const af::dim4& dims) {
+  const auto ndims = dims.ndims();
+  std::vector<int> halideDims(ndims);
+  for (int i = 0; i < ndims; ++i) {
+    halideDims[ndims - 1 - i] = static_cast<int>(dims.dims[i]);
+  }
+  return halideDims;
+}
+
+af::dim4 halideToAfDims(const Halide::Buffer<void>& buffer) {
+  const int nDims = buffer.dimensions();
+  if (nDims > 4) {
+    throw std::invalid_argument(
+        "getDims: Halide buffer has greater than 4 dimensions");
+  }
+  af::dim4 out(1, 1, 1, 1); // initialize so unfilled dims are 1, not 0
+  for (size_t i = 0; i < nDims; ++i) {
+    // Halide can have size zero along a dim --> convert to size 1 for AF
+    auto size = static_cast<dim_t>(buffer.dim(i).extent());
+    out[nDims - 1 - i] = size == 0 ? 1 : size;
+  }
+  return out;
+}
+
+af::dtype halideRuntimeTypeToAfType(halide_type_t type) {
```

Comment:
Halide supports `UInt(64)`, `Int(64)`, and `Float(64)`, as well as 8 (uint/int specific), 16, and 32 bit versions. If flashlight's assumption is that Halide AOT pipelines only use 32 bit numeric types, this likely should be documented somewhere.

Reply:
Fixed to use 2, 4, and 8 byte types - thanks!